File: src/UAFLib/js/CAR.js
function CAR(path, version) {
    //CArchive.call(this, path, version);
    //Object.setPrototypeOf(this, CArchive.prototype);

    if (path == null) {
        this.MFCSerializer = new UAFLib.MFCSerializer();
    } else {
        this.MFCSerializer = new UAFLib.MFCSerializer(path);
    }
    this.MFCSerializer.debug = false;
    this.isStoring = false;
    this.version = version;

    this.m_compressType = 0; //char
    this.stringIndex = { };
    this.m_stringArray = [];
    this.m_nextIndex = 0; //unsigned int 
    this.m_dData = []; // Allocated at run time on heap.  We like
    this.m_cData = []; // to keep these large items off the stack.
    this.m_codes = [];
    this.m_w = 0; //unsigned __int16  
    this.m_bufferIndex = 0; // unsigned int 
    this.m_buffer = [0,0,0,0,0,0,0,0,0,0,0,0,0]; //unsigned int  // 52 bytes  
    this.m_stkLen = 0; // unsigned int 
    this.m_OC = 0; // unsigned __int16 
    this.m_numCode = 0; // unsigned int 
    this.m_C; //unsigned char 


}


CAR.prototype.Compress = function (compression) {
    this.m_compressType = 0;
    if (compression) {
        this.m_codes = new CODES();
        this.m_w = UAFUtil.ByteFromHexString("0xffff");
        this.m_bufferIndex = 0;
        this.m_stkLen = 0;
        this.m_OC = UAFUtil.ByteFromHexString("0xffff");
        this.m_numCode = 256;
        if (this.IsStoring()) {
            this.writeChar(2); // Compression type in clear
            var bufferLen = this.m_buffer.length; for (i = 0; i < bufferLen; i++) { this.m_buffer[i] = 0; };  //takes the place of memset(m_buffer,0,sizeof(m_buffer));
            this.m_compressType = 2;
            this.m_cData = new CDATA(); // Compression Data
        }
        else {
            this.m_dData = new DDATA(); // Decompression Data
            this.m_compressType = this.readChar();
        };
    }
    else {
        if (this.m_cData != null) delete this.m_cData;
        if (this.m_dData != null) delete this.m_dData;
        if (this.m_codes != null) delete this.m_codes;
        this.m_cData = null;
        this.m_dData = null;
        this.m_codes = null;
    }
}

CAR.prototype.readString = function () {
    Globals.debug("In CAR::readString");

    if (this.m_compressType == 0) {
        return this.MFCSerializer.readString();
    }
    var index; // unsigned int
    index = this.MFCSerializer.readInt();
    Globals.debug("index:" + index);
    if (index == 0) { // A new string
        //    *(CArchive *)this >> str;
        var len;
        var temp = null;
        len = this.MFCSerializer.readInt();
        if (len > 1000000) {
            throw (UAFUtil.ByteFromHexString("0x23"));
        };
        if (len == 0) {
            str = "";
        }
        else {
            temp = [];
            if (temp == null) {
                MsgBoxError("Memory exhausted");
                str = "";
            }
            else {
                temp = this.decompress(temp, len);
                //temp[len] = 0;   // PORT NOTE:  null termination not necessary
                str = "" + temp;
                if (this.m_compressType > 1) {
                    if (len != temp.length) {
                        //My_free(temp);
                        return this;
                    };
                };
            };
        };
        this.m_stringArray.SetAtGrow(m_nextIndex, str);
        this.m_nextIndex++;
        //if (temp != null) My_free(temp);
        return this;
    }
    if (index >= this.m_stringArray.length) {
        throw UAFUtil.ByteFromHexString("0x23");
    }
    str = this.m_stringArray[index];
    return this;


    return this.MFCSerializer.readString();
}

CAR.prototype.IsStoring = function () {
    return this.MFCSerializer.IsStoring();
}

CAR.prototype.IsLoading = function () {
    return !this.MFCSerializer.IsStoring();
}

CAR.prototype.readInt = function () {
    if (this.m_compressType == 0) return this.MFCSerializer.readInt32(); //20121208 PRS
    else return this.decompress(4);
}

CAR.prototype.decompress = function (n) {
    return this.MFCSerializer.decompress(n);

/*    var NC = 0;  //unsigned __int16 
    var code = 0;    //unsigned int 
    var theN = 0; //unsigned __int16 
    for (; n > 0; n--, chars++) {
        if (this.m_stkLen != 0) {
        chars = this.m_dData.m_stack[--m_stkLen];
            continue;
        };
        if (this.m_bufferIndex == 0) {
            var bufferLength = 13;  // PORT NOTE: Hard coded since JavaScript has dynamically sized arrays
            if (this.readBytes(this.m_buffer) != bufferLength) return;
        };
        // 
        theN = (unsigned __int16) (((* ((unsigned int *) (((char *)this.m_buffer) + (this.m_bufferIndex >> 3))))>> (m_bufferIndex & 7))& 0x1fff);
        m_bufferIndex = (m_bufferIndex + 13) % 416;
        while (theN == 8190) // RESET
        {
            m_OC = 0xffff;
            m_numCode = 256;
            if (m_bufferIndex == 0) {
                if (Read(m_buffer, 52) != 52) return;
            };
            theN = (unsigned __int16) (((* ((unsigned int *) (((char *)m_buffer) +(m_bufferIndex >> 3))))>> (m_bufferIndex & 7))& 0x1fff);
            m_bufferIndex = (m_bufferIndex + 13) % 416;
        };
        if (m_OC == 0xffff) {
            m_OC = theN;
      * chars = theN;
            m_C = theN;
            continue;
        };
        NC = theN;
        if (NC > 255) {
            if (NC >= m_numCode) {
                if (NC == 8191) return;
                // Expand OC+C into stack;
                m_dData -> m_stack[m_stkLen++] = m_C;
                code = m_OC;
            }
            else {
                // Expand NC onto stack.
                code = NC;
            };
            while (code > 255) {
                ASSERT(code < m_numCode);
                m_dData -> m_stack[m_stkLen++] = m_dData -> m_postfix[code];
                code = m_dData -> m_prefix[code];
            };
            m_dData -> m_stack[m_stkLen++] = (unsigned char) code;
            m_C = m_dData -> m_stack[m_stkLen - 1];
        }
        else {
            m_C = (unsigned char) NC;
            m_dData -> m_stack[m_stkLen++]=m_C;
        };
        m_dData -> m_prefix[m_numCode] = m_OC;
        m_dData -> m_postfix[m_numCode++] = m_C;
        m_OC = NC;
    * chars = m_dData -> m_stack[--m_stkLen];
    };*/
    throw "todo CAR::decompressInt";
}

CAR.prototype.writeInt = function (val) {
    this.MFCSerializer.writeUInt32(val);
}

CAR.prototype.readByte = function () {
    return this.MFCSerializer.readByte();
}

CAR.prototype.writeByte = function (val) {
    this.MFCSerializer.writeByte(val);
}

CAR.prototype.readChar = function () {
    return this.readByte();
}

CAR.prototype.writeChar = function (val) {
    this.writeByte(val);
}

CAR.prototype.writeObject = function (obj) {
    obj.Serialize(this, this.version);
}

CAR.prototype.readString = function () {
    return this.MFCSerializer.readString();
}

CAR.prototype.writeString = function (val) {
    this.MFCSerializer.writeString(val);
}

CAR.prototype.readBool = function () {
    return this.readInt() == 1;
}

CAR.prototype.writeBool = function (val) {
    this.MFCSerializer.writeBool(val);
}

CAR.prototype.readDouble = function () {
    return this.MFCSerializer.readDouble();
}

CAR.prototype.writeDouble = function (val) {
    this.MFCSerializer.writeDouble(val);
}

CAR.prototype.readDWORD = function () {
    return this.readInt();
}

CAR.prototype.readWORD = function () {
    return this.MFCSerializer.readInt16();
}

CAR.prototype.readFloat = function () {
    return this.MFCSerializer.readFloat();
}


CAR.prototype.Close = function () {
    return this.MFCSerializer.Close();
}

File: UAFCArchive/CAR.cpp
#include "pch.h"
#include "CAR.h"



//CAR::CAR(CFile *pFile, UINT nMode) :  CArchive(pFile, nMode)
CAR::CAR(CFile* pFile, UINT nMode) : ar(pFile, nMode)
{
    m_compressType = 0; // No compression yet.
    m_nextIndex = 1;
    m_cData = NULL;
    m_dData = NULL;
    m_codes = NULL;
    m_w = 0;
    m_bufferIndex = 0;
    memset(m_buffer, 0, sizeof(m_buffer));
    m_stkLen = 0;
    m_OC = 0;
    m_numCode = 0;
    m_C = 0;
}


CAR::~CAR(void)
{
    // Make sure we flush the output buffer
    //Flush();
    if (m_cData != NULL) delete m_cData;
    if (m_dData != NULL) delete m_dData;
    if (m_codes != NULL) delete m_codes;
}

void CAR::compress(const char* chars, int n)
{
    //  static unsigned int buffer[13];
    //  static unsigned int bufferIndex=0;
    //  static unsigned _int16 w=0xffff;
    unsigned __int16 cd;
    unsigned char c;
    for (; n > 0; n--, chars++)
    {
        if (m_codes->Full())
        {
            // And output the old code from w.
            *((unsigned int*)(((char*)m_buffer) + (m_bufferIndex >> 3))) |= m_w << (m_bufferIndex & 7);
            m_bufferIndex = (m_bufferIndex + 13) % 416;
            if (m_bufferIndex == 0)
            {
                Write(m_buffer, 52);
                memset(m_buffer, 0, 52);
            };
            m_w = 0xffff;
            m_codes->Clear();
            // And output the RESET code
            *((unsigned int*)(((char*)m_buffer) + (m_bufferIndex >> 3))) |= 8190 << (m_bufferIndex & 7);
            m_bufferIndex = (m_bufferIndex + 13) % 416;
            if (m_bufferIndex == 0)
            {
                Write(m_buffer, 52);
                memset(m_buffer, 0, 52);
            };
        };
        c = *chars;
        if (m_w == 0xffff)
        { // Single characters are in code list by definition.
            m_w = c;
            continue;
        };
        if (((cd = m_codes->FindCode((m_w << 8) | c)) != 0) /*&& (c<256) What was this for????*/)
        {
            // We found the new, larger string in the table.
            m_w = cd;
            if (m_w > 8189) // 8191 reserved for EOF, 8190 for RESET
            {
                _exit(1);
            };
            continue;
        };
        // We want to add the new, larger string to the table.
        m_codes->AddCode((m_w << 8) | c);
        // And output the old string from m_w.
        *((unsigned int*)(((char*)m_buffer) + (m_bufferIndex >> 3))) |= m_w << (m_bufferIndex & 7);
        m_bufferIndex = (m_bufferIndex + 13) % 416;
        if (m_bufferIndex == 0)
        {
            Write(m_buffer, sizeof(m_buffer));
            memset(m_buffer, 0, sizeof(m_buffer));
        };
        m_w = c;
    };
}

void CAR::decompress(char* chars, int n)
{
    // static unsigned _int16 OC=0xffff;
    // static unsigned _int16 prefix[8192];
    // static unsigned char   postfix[8192];
    // static unsigned int numCode=256;
    unsigned __int16 NC;
    // static unsigned char stack[1000];
    // static unsigned char C;
    // static unsigned int stkLen=0;
    unsigned int code;
    // static unsigned int buffer[13];
    // static unsigned int bufferIndex=0;
    unsigned __int16 theN;
    for (; n > 0; n--, chars++)
    {
        if (m_stkLen != 0)
        {
            *chars = m_dData->m_stack[--m_stkLen];
            continue;
        };
        if (m_bufferIndex == 0)
        {
            if (Read(m_buffer, sizeof(m_buffer)) != sizeof(m_buffer)) return;
        };
        theN = (unsigned __int16)(((*((unsigned int*)(((char*)m_buffer) + (m_bufferIndex >> 3)))) >> (m_bufferIndex & 7)) & 0x1fff);
        m_bufferIndex = (m_bufferIndex + 13) % 416;
        while (theN == 8190) // RESET
        {
            m_OC = 0xffff;
            m_numCode = 256;
            if (m_bufferIndex == 0)
            {
                if (Read(m_buffer, 52) != 52) return;
            };
            theN = (unsigned __int16)(((*((unsigned int*)(((char*)m_buffer) + (m_bufferIndex >> 3)))) >> (m_bufferIndex & 7)) & 0x1fff);
            m_bufferIndex = (m_bufferIndex + 13) % 416;
        };
        if (m_OC == 0xffff)
        {
            m_OC = theN;
            *chars = theN;
            m_C = theN;
            continue;
        };
        NC = theN;
        if (NC > 255)
        {
            if (NC >= m_numCode)
            {
                if (NC == 8191) return;
                // Expand OC+C into stack;
                m_dData->m_stack[m_stkLen++] = m_C;
                code = m_OC;
            }
            else
            {
                // Expand NC onto stack.
                code = NC;
            };
            while (code > 255)
            {
                ASSERT(code < m_numCode);
                m_dData->m_stack[m_stkLen++] = m_dData->m_postfix[code];
                code = m_dData->m_prefix[code];
            };
            m_dData->m_stack[m_stkLen++] = (unsigned char)code;
            m_C = m_dData->m_stack[m_stkLen - 1];
        }
        else
        {
            m_C = (unsigned char)NC;
            m_dData->m_stack[m_stkLen++] = m_C;
        };
        m_dData->m_prefix[m_numCode] = m_OC;
        m_dData->m_postfix[m_numCode++] = m_C;
        m_OC = NC;
        *chars = m_dData->m_stack[--m_stkLen];
    };
}



CAR::CODES::CODES(void)
{
    Clear();
    m_numMisses = 0;
    m_numCodes = 0;
}

CAR::CODES::~CODES(void)
{
}

void CAR::CODES::Clear(void)
{
    memset(m_hashTable, 0, sizeof(m_hashTable));
    memset(m_codes, 0xff, 256 * sizeof(unsigned int));
    m_nextCode = 256;
}

unsigned __int16 CAR::CODES::AddCode(unsigned int code)
{
    unsigned int hashcode, incr;
    m_numCodes++;
    ASSERT(m_nextCode <= 8189);
    incr = hashcode = code % (sizeof(m_hashTable) / sizeof(m_hashTable[0]));
    if (incr == 0) incr = 1;
    while (m_hashTable[hashcode] != 0)
    {
        m_numMisses++;
        hashcode = (hashcode + incr) % (sizeof(m_hashTable) / sizeof(m_hashTable[0]));
    };
    m_hashTable[hashcode] = m_nextCode;
    m_codes[m_nextCode] = code;
    return m_nextCode++;
}

unsigned __int16 CAR::CODES::FindCode(unsigned int code)
{
    unsigned int hashcode, incr;
    incr = hashcode = code % (sizeof(m_hashTable) / sizeof(m_hashTable[0]));
    if (incr == 0) incr = 1;
    while (m_hashTable[hashcode] != 0)
    {
        if (m_codes[m_hashTable[hashcode]] == code) return m_hashTable[hashcode];
        hashcode = (hashcode + incr) % (sizeof(m_hashTable) / sizeof(m_hashTable[0]));
    };
    return 0;
}

void CAR::CODES::HashDump(FILE* f)
{
    int code, stklen;
    int stk[1000];
    for (int i = 256; i < m_nextCode; i++)
    {
        code = i;
        fprintf(f, "%4d ", code);
        stklen = 0;
        while (code > 255)
        {
            code = m_codes[code];
            stk[stklen++] = code & 0xff;
            code = code >> 8;
        };
        stk[stklen++] = code;
        while (stklen > 0)
        {
            fprintf(f, "%3d ", stk[--stklen]);
        };
        fprintf(f, "\n");
    };
}
